cmd/main.go

```diff
@@ -84,6 +91,71 @@ func main() {
 		l.Fatalf("Enclave terminated: %v", err)
 	}
 
-	// Block on this read forever.
-	<-make(chan struct{})
+	// Nitriding supports two ways of starting the enclave application:
+	//
+	// 1) Nitriding spawns the enclave application itself, and waits for it
+	//    to terminate.
+	//
+	// 2) The enclave application is started by a shell script (which also
+	//    starts nitriding).  In this case, we simply block forever.
+	if appCmd != "" {
+		f := func(s string) {
+			l.Printf("Application says: %s", s)
+		}
+		runAppCommand(appCmd, f, f)
+	} else {
+		// Block forever.
+		<-make(chan struct{})
+	}
+	l.Println("Exiting nitriding.")
+}
+
+// runAppCommand (i) runs the given command, (ii) waits until the command
+// finished execution, and (iii) in the meanwhile prints the command's stdout
+// and stderr.
+func runAppCommand(appCmd string, stdoutFunc, stderrFunc func(string)) {
+	l.Printf("Invoking the enclave application.")
+	args := strings.Split(appCmd, " ")
+	cmd := exec.Command(args[0], args[1:]...)
+
+	// Print the enclave application's stderr.
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		l.Fatalf("Failed to obtain stderr pipe for enclave application: %v", err)
+	}
+	go printOutput(stderr, stderrFunc, "stderr")
+
+	// Print the enclave application's stdout.
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		l.Fatalf("Failed to obtain stdout pipe for enclave application: %v", err)
+	}
+	go printOutput(stdout, stdoutFunc, "stdout")
+
+	if err := cmd.Start(); err != nil {
+		l.Fatalf("Failed to start enclave application: %v", err)
+	}
+
+	if err := cmd.Wait(); err != nil {
+		l.Fatalf("Enclave application exited with non-0 exit code: %v", err)
+	}
+	l.Println("Enclave application exited.")
+}
+
+// printOutput continuously reads from the given Reader until an EOF occurs.
+// Each newly read line is passed to the given function f.
+func printOutput(readCloser io.ReadCloser, f func(string), output string) {
```

Comment:
This could be `forwardOutput` to reflect that it's just forwarding to a callback, which may not actually print anything.

Reply:
Separately, does taking an `io.ReadCloser` here do anything? I don't see an explicit `close` call, and presumedly we want to keep the underlying pipe open?

Reply:
The documentation of [`StderrPipe`](https://pkg.go.dev/os/exec#Cmd.StderrPipe) (and `StdoutPipe`) says:
> Wait will close the pipe after seeing the command exit, so most callers need not close the pipe themselves.

Reply:
> This could be `forwardOutput` to reflect that it's just forwarding to a callback, which may not actually print anything.

Addressed in caf63c5da1844f562215b51c21ca3b8316532bd8.

Reply:
I see. Thanks for explaining.